chainer/link_hooks/spectral_normalization.py

```diff
@@ -0,0 +1,216 @@
+    be added to :func:`chainer.links.Linear`, :func:`chainer.links.EmbedID`,
+    :func:`chainer.links.Convolution2D`, :func:`chainer.links.ConvolutionND`,
+    :func:`chainer.links.Deconvolution2D`,
+    and :func:`chainer.links.DeconvolutionND`. However, you can use this to
+    other links like RNNs by specifying ``weight_name``.
+
+    .. math::
+         \mathbf{W} &:=& \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \text{where,} \ \sigma(\mathbf{W}) &:=&
+         \max_{\mathbf{h}: \mathbf{h} \ne 0}
+         \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+         = \max_{\|\mathbf{h}\|_2 \le 1} \|\mathbf{W}\mathbf{h}\|_2
+
+    See: T. Miyato et. al., `Spectral Normalization for Generative Adversarial
+    Networks <https://arxiv.org/abs/1802.05957>`_
+
+    Args:
+        n_power_iteration (int): Number of power iteration.
+            The default value is 1.
+        eps (int): Numerical stability in norm calculation.
+            The default value is 1e-12.
+        use_gamma (bool): If ``True``, weight scaling parameter gamma which is
+            initialized by initial weight's max singular value is introduced.
+        factor (float, None): Scaling parameter to divide maximum singular value.
+            The default value is 1.0.
+        weight_name (str): Link's weight name to apply this hook. The default
+            value is 'W'.
+        name (str or None): Name of this hook. The default value is
+            'SpectralNormalization'.
+
+    Attributes:
+        vector_name (str): Name of the approximate first left singular vector
+            registered in the target link.
+            the target link.
+        axis (int): Axis of weight represents the numbef of output
+            feature maps or output units (``out_channels`` and
+            ``out_size``, respectively).
+
+    .. admonition:: Example
+
+        >>> x = np.random.uniform(-1, 1, (10, 3, 32, 32)).astype(np.float32)
+        >>> layer = L.Convolution2D(3, 5, 3, 1, 1)
+        >>> hook = SpectralNormalization()
+        >>> layer.add_hook(hook)
+        >>> y = layer(x)
+
+    """
+
+    name = 'SpectralNormalization'
+
+    def __init__(self, n_power_iteration=1, eps=1e-12, use_gamma=False,
+                 factor=None, weight_name='W', name=None):
+        assert n_power_iteration > 0
+        self.n_power_iteration = n_power_iteration
+        self.eps = eps
+        self.use_gamma = use_gamma
+        self.factor = factor
+        self.weight_name = weight_name
+        self.vector_name = weight_name + '_u'
+        self._initialied = False
+        self.axis = 0
+
+        if name is not None:
+            self.name = name
+
+    def added(self, link):
+        if isinstance(
+            link, (
+                L.Deconvolution1D, L.Deconvolution2D,
+                L.Deconvolution3D, L.DeconvolutionND)):
+            self.axis = 1
+        if link.W.array is not None:
+            self._prepare_parameters(link)
+
+    def deleted(self, link):
```

Comment:
I have some concerns here. 
When this method is called after training (before evaluation), assigning the latest weight w/o normalization is not harmful to me because the weight is updated and weight's values are different from those used in the previous forward method in despite of whether normalized or not. 
However, when this is called after evaluation, assigning the unnormalized weight means that the values are not the same as the ones used in the previous iterations because in evaluation weight is normalized. This behavior is not desirable to me (and some users).
So, I choose to assign the weight after normalization here. What do you think?